python/llm/test/inference_gpu/test_transformers_api_layernorm.py

```diff
@@ -0,0 +1,119 @@
+            model = model.to(device)
+            for layer_name, layer_module in model.named_modules():
+                if layer_name == layer_before_LayerNorm:
+                    layer_module.register_forward_hook(
+                        lambda module, input, output, layer_name=layer_name: pre_forward_hook(module, input,
+                                                                                            output, layer_name))
+                if layer_name == LayerNorm_layer:
+                    layer_module.register_forward_hook(
+                        lambda module, input, output, layer_name=layer_name: forward_hook(module, input,
+                                                                                        output, layer_name))
+            logits_base_model = (model(input_ids)).logits
+            # the list `layer_output` has only one element.
+            layer_tensor = self.layer_outputs.pop()
+            model.to('cpu')
+ 
+            opt_model = Model.from_pretrained(model_path,
+                                            load_in_4bit=True,
+                                            optimize_model=True,
+                                            trust_remote_code=True)
+            opt_model = opt_model.to(device)
+ 
+ 
+            def replace_forward_hook(module, input, output, layer_name):
+                output = self.pre_layer_outputs[0]
+                return output
+ 
+            for layer_name, layer_module in opt_model.named_modules():
+                if layer_name == layer_before_LayerNorm:
+                    layer_module.register_forward_hook(
+                        lambda module, input, output, layer_name=layer_name: replace_forward_hook(module, input,
+                                                                                                output, layer_name))
+                if layer_name == LayerNorm_layer:
+                    layer_module.register_forward_hook(
+                        lambda module, input, output, layer_name=layer_name: forward_hook(module, input,
+                                                                                        output, layer_name))
+            logits_optimized_model = (opt_model(input_ids)).logits
+            # the list `layer_output` has only one element.
+            opt_layer_tensor = self.layer_outputs[0]
+            opt_model.to('cpu')
+ 
+ 
+            LayerNorm_output_diff = []
+            for i, (t1, t2) in enumerate(zip(layer_tensor, opt_layer_tensor)):
+                if t1 is not None and t2 is not None:
+                    if isinstance(t1, torch.Tensor) and isinstance(t2, torch.Tensor):
+                        LayerNorm_output_diff.append(t1 - t2)
+                    else:
+                        # 'past_key_value'is of type tuple as default.
+                        for i, (t3, t4) in enumerate(zip(t1, t2)):
+                            LayerNorm_output_diff.append(t3 - t4)
```

Comment:
Can we remove this? since for layernorm it is always a tensor. past kv is for attention only?
May also remove in mlp